plugins/services/src/js/filters/__tests__/GenericObject-test.ts

```diff
@@ -0,0 +1,50 @@
+import { filterByObject } from "../GenericObject";
+
+describe("GenericObject", () => {
+  describe("#filterByObject", () => {
+    it("can filter by string", () => {
+      expect(filterByObject({ foo: "bar" }, { foo: "baz" })).toEqual(false);
+      expect(filterByObject({ foo: "bar" }, { foo: "bar" })).toEqual(true);
+    });
+    it("can filter by number", () => {
+      expect(filterByObject({ foo: 100 }, { foo: 0 })).toEqual(false);
+      expect(filterByObject({ foo: 100 }, { foo: 100 })).toEqual(true);
+    });
+    it("can filter by boolean", () => {
+      expect(filterByObject({ foo: true }, { foo: false })).toEqual(false);
+      expect(filterByObject({ foo: true }, { foo: true })).toEqual(true);
+    });
+    it("can filter for undefined", () => {
+      expect(filterByObject({ foo: true }, { foo: undefined })).toEqual(false);
+      expect(filterByObject({ foo: true }, { bar: undefined })).toEqual(true);
```

Comment:
i admire the complexity of `filterByObject` but have no idea what it does.

i imagine jest's `objectContaining`?

in particular i don't understand this very test. why would i want this to be true, if my `foo` does not contain a `bar`? i would want the key to be present, even if its value is undefined.

Currently we only ever search for `{ quota: { enforced: true } }`, right?

Reply:
if it wasn't for that case we could do something like: 

```typescript
type Rec = Record<string, unknown>;

export function filterByObject(value: Rec, filterObj: Rec): boolean {
  const keys = Object.keys(value || {});

  return Object.keys(filterObj || {}).every(k =>
    keys.includes(k) ? matchValue(value[k], filterObj[k]) : false
  );
}

const matchValue = (a: unknown, b: unknown) => {
  if (isRecord(a) && isRecord(b)) {
    return filterByObject(a, b);
  }
  return a === b;
};
const isRecord = (a: unknown): a is Rec => typeof a === "object" && a !== null;
```

this makes all but the `undefined` test pass.

Reply:
I added `filterByObject` so we could easily send a filter object to the graphQL query. We currently only need it to get groups with quota enforced for the overview page, but that is temporary for MVP, we would want to show all groups in that view in the future.

And I didn’t want to have a single Boolean on a group query because that would of been too specific. So I tried to write filterByObject as something more generic that can be expanded upon. But also not to invest too much time now since we only needed to filter by a single sub property value.

Maybe I overdid it there but I was trying to balance current feature requirements with the spirit of graphql and future needs.

Reply:
i added a refactor-commit changing the behaviour in the mentioned case because i think that if a key with an `undefined` value has significance in

```
filterByObject({ foo: true }, { foo: undefined })
```

it should also have significance in

```
(filterByObject({ foo: true }, { bar: undefined })
```

If you don't like it, feel free to throw away that commit again.